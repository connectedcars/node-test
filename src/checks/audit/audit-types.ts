// actions: [],
// advisories: {},
// muted: [],
// metadata: {
//   vulnerabilities: {
//     info: 0,
//     low: 0,
//     moderate: 0,
//     high: 0,
//     critical: 0
//   },
//   dependencies: 2364,
//   devDependencies: 2606,
//   optionalDependencies: 34,
//   totalDependencies: 4982
// },
// runId: 'd558cef6-7061-454e-89d3-ee85181d1dd6'

export interface AuditData {
  actions: Action[]
  advisories: Advisories
  metadata: AuditMetadata
  runId: string
}

export interface Action {
  action: string
  module: string
  depth: number
  target: string
}

interface AuditMetadata {
  vulnerabilities: Vulnerabilities
  dependencies: number
  devDependencies: number
  optionalDependencies: number
  totalDependencies: number
}

interface Advisories {
  [key: string]: Advisory
}

interface Advisory {
  findings: Finding[]
  id: number
  url: string // check this
  module_name: string
  title: string
  overview: string
  severity: string
  vulnerable_versions: string
  recommendation: string
}

interface Finding {
  paths: string[]
  dev: string
}

export interface Vulnerabilities {
  [key: string]: number
  // info: number
  // low: number
  // moderate: number
  // high: number
  // critical: number
  // all: number
}
