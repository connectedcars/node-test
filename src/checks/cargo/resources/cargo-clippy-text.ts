import { CargoMessage } from '../cargo-types'

export const cargoClippyFailedOutput: CargoMessage[] = [
  {
    reason: 'compiler-message',
    package_id: 'can 0.1.0 (path+file:///app/can)',
    target: {
      kind: ['lib'],
      crate_types: ['lib'],
      name: 'can',
      src_path: '/app/can/src/lib.rs',
      edition: '2018',
      doctest: true
    },
    message: {
      rendered:
        'error: this comparison involving the minimum or maximum element for this type contains a case that is always true or always false\n   --> can/src/can_session.rs:391:12\n    |\n391 |         if 100 > i32::MAX {}\n    |            ^^^^^^^^^^^^^^\n    |\n    = note: `-D clippy::absurd-extreme-comparisons` implied by `-D clippy::all`\n    = help: because `i32::MAX` is the maximum value for this type, this comparison is always false\n    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#absurd_extreme_comparisons\n\n',
      children: [
        {
          children: [],
          code: null,
          level: 'note',
          message: '`-D clippy::absurd-extreme-comparisons` implied by `-D clippy::all`',
          rendered: null,
          spans: []
        },
        {
          children: [],
          code: null,
          level: 'help',
          message: 'because `i32::MAX` is the maximum value for this type, this comparison is always false',
          rendered: null,
          spans: []
        },
        {
          children: [],
          code: null,
          level: 'help',
          message:
            'for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#absurd_extreme_comparisons',
          rendered: null,
          spans: []
        }
      ],
      code: {
        code: 'clippy::absurd_extreme_comparisons',
        explanation: null
      },
      level: 'error',
      message:
        'this comparison involving the minimum or maximum element for this type contains a case that is always true or always false',
      spans: [
        {
          byte_end: 13338,
          byte_start: 13324,
          column_end: 26,
          column_start: 12,
          expansion: null,
          file_name: 'can/src/can_session.rs',
          is_primary: true,
          label: null,
          line_end: 391,
          line_start: 391,
          suggested_replacement: null,
          suggestion_applicability: null,
          text: [
            {
              highlight_end: 26,
              highlight_start: 12,
              text: '        if 100 > i32::MAX {}'
            }
          ]
        }
      ]
    }
  },
  {
    reason: 'compiler-message',
    package_id: 'can 0.1.0 (path+file:///app/can)',
    target: {
      kind: ['lib'],
      crate_types: ['lib'],
      name: 'can',
      src_path: '/app/can/src/lib.rs',
      edition: '2018',
      doctest: true
    },
    message: {
      rendered: 'error: aborting due to previous error\n\n',
      children: [],
      code: null,
      level: 'error',
      message: 'aborting due to previous error',
      spans: []
    }
  }
]
