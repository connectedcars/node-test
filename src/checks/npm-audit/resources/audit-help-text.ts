export const auditOneVulnerability = `{ "actions": [{ "action": "install", "module": "knex", "target": "3.0.0", "isMajor": true, "resolves": [{ "id": 525, "path": "knex>liftoff>findup-sync>micromatch>braces>expand-range>fill-range>randomatic", "dev": false, "optional": false }] }], "advisories": { "525": { "id": 525, "created": "2017-09-08T18:07:02.061Z", "updated": "2017-09-22T16:26:08.422Z", "deleted": null, "title": "Regular Expression Denial of Service", "found_by": { "name": "nobody" }, "reported_by": { "name": "testdata" }, "module_name": "knex", "cves": [ "CVE-2017-16112" ], "vulnerable_versions": "<3.0.0", "patched_versions": ">=3.0.0", "overview": "something here", "recommendation": "Please update to version 3.0.0 or greater", "references": "- https://github.com/salesforce/tough-cookie/issues/92", "access": "public", "severity": "high", "cwe": "CWE-400" } }, "muted": [], "metadata": { "vulnerabilities": { "low": 0, "moderate": 0, "high": 1, "critical": 0 }, "dependencies": 375, "devDependencies": 466, "optionalDependencies": 87, "totalDependencies": 918 } }`

export const auditNoVulnerabilities = `{ "actions": [], "advisories": {}, "muted": [], "metadata": { "vulnerabilities": { "info": 0, "low": 0, "moderate": 0, "high": 0, "critical": 0 }, "dependencies": 2364, "devDependencies": 2606, "optionalDependencies": 34, "totalDependencies": 4982 }, "runId": "d558cef6-7061-454e-89d3-ee85181d1dd6" }`

export const auditMultiVulnerabilities = `{ "actions": [], "advisories": { "146": { "findings": [ { "version": "1.9.2", "paths": [ "mocha>growl" ], "dev": true, "optional": false, "bundled": false } ], "id": 146, "created": "2016-09-06T12:49:40.000Z", "updated": "2018-03-02T21:07:28.071Z", "deleted": null, "title": "Command Injection", "found_by": { "name": "Cristian-Alexandru Staicu" }, "reported_by": { "name": "Cristian-Alexandru Staicu" }, "module_name": "growl", "cves": [ "CVE-2017-16042" ], "vulnerable_versions": "<1.10.2", "patched_versions": ">=1.10.2", "overview": "Affected versions of \`growl\` do not properly sanitize input prior to passing it into a shell command, allowing for arbitrary command execution.", "recommendation": "Update to version 1.10.2 or later.", "references": "[Issue #60](https://github.com/tj/node-growl/issues/60)\\n[PR #61](https://github.com/tj/node-growl/pull/61)", "access": "public", "severity": "critical", "cwe": "CWE-94", "metadata": { "module_type": "CLI.Library", "exploitability": 5, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/146" }, "157": { "findings": [ { "version": "1.1.7", "paths": [ "nyc>micromatch>braces>expand-range>fill-range>randomatic", "nyc>test-exclude>micromatch>braces>expand-range>fill-range>randomatic" ], "dev": true, "optional": false, "bundled": true } ], "id": 157, "created": "2016-11-09T20:03:19.000Z", "updated": "2018-05-08T15:23:56.190Z", "deleted": null, "title": "Cryptographically Weak PRNG", "found_by": { "name": "Sven Slootweg" }, "reported_by": { "name": "Sven Slootweg" }, "module_name": "randomatic", "cves": [ "CVE-2017-16028" ], "vulnerable_versions": "<3.0.0", "patched_versions": ">=3.0.0", "overview": "Affected versions of \`randomatic\` generate random values using a cryptographically weak psuedo-random number generator. This may result in predictable values instead of random values as intended.\\r\\n\\r\\n", "recommendation": "Update to version 3.0.0 or later.\\r\\n", "references": "- [Commit #4a52695](https://github.com/jonschlinkert/randomatic/commit/4a526959b3a246ae8e4a82f9c182180907227fe1#diff-b9cfc7f2cdf78a7f4b91a753d10865a2)", "access": "public", "severity": "low", "cwe": "CWE-330", "metadata": { "module_type": "Multi.Library", "exploitability": 5, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/157" }, "525": { "findings": [ { "version": "2.3.2", "paths": [ "coveralls>request>tough-cookie" ], "dev": true, "optional": false, "bundled": false } ], "id": 525, "created": "2017-09-08T18:07:02.061Z", "updated": "2018-04-09T00:01:19.079Z", "deleted": null, "title": "Regular Expression Denial of Service", "found_by": { "name": "Cristian-Alexandru Staicu" }, "reported_by": { "name": "Cristian-Alexandru Staicu" }, "module_name": "tough-cookie", "cves": [ "CVE-2017-16112" ], "vulnerable_versions": "<2.3.3", "patched_versions": ">=2.3.3", "overview": "Affected versions of \`tough-cookie\` are susceptible to a regular expression denial of service.\\n\\nThe amplification on this vulnerability is relatively low - it takes around 2 seconds for the engine to execute on a malicious input which is 50,000 characters in length.\\n\\nIf node was compiled using the \`-DHTTP_MAX_HEADER_SIZE\` however, the impact of the vulnerability can be significant, as the primary limitation for the vulnerability is the default max HTTP header length in node.", "recommendation": "Update to version 2.3.3 or later.", "references": "[Issue #92](https://github.com/salesforce/tough-cookie/issues/92)", "access": "public", "severity": "high", "cwe": "CWE-400", "metadata": { "module_type": "Network.Library", "exploitability": 5, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/525" }, "534": { "findings": [ { "version": "2.6.8", "paths": [ "babel-eslint>babel-traverse>debug", "eslint>debug", "mocha>debug", "nyc>istanbul-lib-instrument>babel-template>babel-traverse>debug", "nyc>istanbul-lib-instrument>babel-traverse>debug", "nyc>istanbul-lib-source-maps>debug" ], "dev": true, "optional": false, "bundled": false } ], "id": 534, "created": "2017-09-25T18:55:55.956Z", "updated": "2018-05-16T19:37:43.686Z", "deleted": null, "title": "Regular Expression Denial of Service", "found_by": { "name": "Cristian-Alexandru Staicu" }, "reported_by": { "name": "Cristian-Alexandru Staicu" }, "module_name": "debug", "cves": [ "CVE-2017-16137" ], "vulnerable_versions": "<= 2.6.8 || >= 3.0.0 <= 3.0.1", "patched_versions": ">= 2.6.9 < 3.0.0 || >= 3.1.0", "overview": "Affected versions of \`debug\` are vulnerable to regular expression denial of service when untrusted user input is passed into the \`o\` formatter. \\n\\nAs it takes 50,000 characters to block the event loop for 2 seconds, this issue is a low severity issue.", "recommendation": "Version 2.x.x: Update to version 2.6.9 or later.\\nVersion 3.x.x: Update to version 3.1.0 or later.\\n", "references": "- [Issue #501](https://github.com/visionmedia/debug/issues/501)\\n- [PR #504](https://github.com/visionmedia/debug/pull/504)", "access": "public", "severity": "low", "cwe": "CWE-400", "metadata": { "module_type": "", "exploitability": 5, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/534" }, "566": { "findings": [ { "version": "2.16.3", "paths": [ "coveralls>request>hawk>boom>hoek", "coveralls>request>hawk>cryptiles>boom>hoek", "coveralls>request>hawk>hoek", "coveralls>request>hawk>sntp>hoek" ], "dev": true, "optional": false, "bundled": false } ], "id": 566, "created": "2018-04-20T21:25:58.421Z", "updated": "2018-04-20T21:25:58.421Z", "deleted": null, "title": "Prototype pollution", "found_by": { "name": "HoLyVieR" }, "reported_by": { "name": "HoLyVieR" }, "module_name": "hoek", "cves": [], "vulnerable_versions": "<= 4.2.0 || >= 5.0.0 < 5.0.3", "patched_versions": "> 4.2.0 < 5.0.0 || >= 5.0.3", "overview": "Versions of \`hoek\` prior to 4.2.1 and 5.0.3 are vulnerable to prototype pollution.\\n\\nThe \`merge\` function, and the \`applyToDefaults\` and \`applyToDefaultsWithShallow\` functions which leverage \`merge\` behind the scenes, are vulnerable to a prototype pollution attack when provided an _unvalidated_ payload created from a JSON string containing the \`__proto__\` property.\\n\\nThis can be demonstrated like so:\\n\\n\`\`\`javascript\\nvar Hoek = require('hoek');\\nvar malicious_payload = '{\\"__proto__\\":{\\"oops\\":\\"It works !\\"}}';\\n\\nvar a = {};\\nconsole.log(\\"Before : \\" + a.oops);\\nHoek.merge({}, JSON.parse(malicious_payload));\\nconsole.log(\\"After : \\" + a.oops);\\n\`\`\`\\n\\nThis type of attack can be used to overwrite existing properties causing a potential denial of service.", "recommendation": "Update to version 4.2.1, 5.0.3 or later.", "references": "", "access": "public", "severity": "moderate", "cwe": "CWE-471", "metadata": { "module_type": "", "exploitability": 5, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/566" }, "577": { "findings": [ { "version": "4.17.4", "paths": [ "babel-eslint>babel-traverse>babel-types>lodash", "babel-eslint>babel-traverse>lodash", "babel-eslint>babel-types>lodash", "eslint>inquirer>lodash", "eslint>lodash", "eslint>table>lodash", "nyc>istanbul-lib-instrument>babel-generator>babel-types>lodash", "nyc>istanbul-lib-instrument>babel-generator>lodash", "nyc>istanbul-lib-instrument>babel-template>babel-traverse>babel-types>lodash", "nyc>istanbul-lib-instrument>babel-template>babel-traverse>lodash", "nyc>istanbul-lib-instrument>babel-template>babel-types>lodash", "nyc>istanbul-lib-instrument>babel-template>lodash", "nyc>istanbul-lib-instrument>babel-traverse>babel-types>lodash", "nyc>istanbul-lib-instrument>babel-traverse>lodash", "nyc>istanbul-lib-instrument>babel-types>lodash" ], "dev": true, "optional": false, "bundled": false } ], "id": 577, "created": "2018-04-24T14:27:02.796Z", "updated": "2018-04-24T14:27:13.049Z", "deleted": null, "title": "Prototype Pollution", "found_by": { "name": "Olivier Arteau (HoLyVieR)" }, "reported_by": { "name": "Olivier Arteau (HoLyVieR)" }, "module_name": "lodash", "cves": [ "CVE-2018-3721" ], "vulnerable_versions": "<4.17.5", "patched_versions": ">=4.17.5", "overview": "Versions of \`lodash\` before 4.17.5 are vulnerable to prototype pollution. \\n\\nThe vulnerable functions are 'defaultsDeep', 'merge', and 'mergeWith' which allow a malicious user to modify the prototype of \`Object\` via \`__proto__\` causing the addition or modification of an existing property that will exist on all objects.\\n\\n", "recommendation": "Update to version 4.17.5 or later.", "references": "- [HackerOne Report](https://hackerone.com/reports/310443)", "access": "public", "severity": "low", "cwe": "CWE-471", "metadata": { "module_type": "", "exploitability": 1, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/577" }, "598": { "findings": [ { "version": "0.4.3", "paths": [ "coveralls>request>tunnel-agent" ], "dev": true, "optional": false, "bundled": false } ], "id": 598, "created": "2018-04-24T20:30:16.099Z", "updated": "2018-04-24T20:31:15.816Z", "deleted": null, "title": "Memory Exposure", "found_by": { "name": "Сковорода Никита Андреевич" }, "reported_by": { "name": "Сковорода Никита Андреевич" }, "module_name": "tunnel-agent", "cves": [], "vulnerable_versions": "<0.6.0", "patched_versions": ">=0.6.0", "overview": "Versions of \`tunnel-agent\` before 0.6.0 are vulnerable to memory exposure.\\n\\nThis is exploitable if user supplied input is provided to the auth value and is a number.\\n\\nProof-of-concept:\\n\`\`\`js\\nrequire('request')({\\n  method: 'GET',\\n  uri: 'http://www.example.com',\\n  tunnel: true,\\n  proxy:{\\n    protocol: 'http:',\\n    host:'127.0.0.1',\\n    port:8080,\\n    auth:USERSUPPLIEDINPUT // number\\n  }\\n});\\n\`\`\`", "recommendation": "Update to version 0.6.0 or later.", "references": "- [GitHub Commit #9ca95ec](https://github.com/request/tunnel-agent/commit/9ca95ec7219daface8a6fc2674000653de0922c0)\\n- [Proof of Concept](https://gist.github.com/ChALkeR/fd6b2c445834244e7d440a043f9d2ff4)", "access": "public", "severity": "moderate", "cwe": "CWE-20", "metadata": { "module_type": "", "exploitability": 3, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/598" }, "606": { "findings": [ { "version": "1.13.1", "paths": [ "coveralls>request>http-signature>sshpk" ], "dev": true, "optional": false, "bundled": false } ], "id": 606, "created": "2018-04-24T22:25:08.333Z", "updated": "2018-04-24T22:25:13.388Z", "deleted": null, "title": "Regular Expression Denial of Service", "found_by": { "name": "Сковорода Никита Андреевич" }, "reported_by": { "name": "Сковорода Никита Андреевич" }, "module_name": "sshpk", "cves": [], "vulnerable_versions": "<1.14.1", "patched_versions": ">=1.14.1", "overview": "Versions of \`sshpk\` before 1.14.1 are vulnerable to regular expression denial of service when parsing crafted invalid public keys.", "recommendation": "Update to version 1.14.1 or later.", "references": "- https://github.com/joyent/node-sshpk/blob/v1.13.1/lib/formats/ssh.js#L17\\n- [HackerOne Report[(https://hackerone.com/reports/319593)", "access": "public", "severity": "high", "cwe": "CWE-400", "metadata": { "module_type": "", "exploitability": 5, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/606" }, "664": { "findings": [ { "version": "0.0.5", "paths": [ "coveralls>request>stringstream" ], "dev": true, "optional": false, "bundled": false } ], "id": 664, "created": "2018-05-16T19:39:37.463Z", "updated": "2018-05-22T15:03:12.999Z", "deleted": null, "title": "Out-of-bounds Read", "found_by": { "name": "Сковорода Никита Андреевич" }, "reported_by": { "name": "Сковорода Никита Андреевич" }, "module_name": "stringstream", "cves": [], "vulnerable_versions": "<=0.0.5", "patched_versions": ">=0.0.6", "overview": "All versions of \`stringstream\` are vulnerable to out-of-bounds read as it allocates uninitialized Buffers when number is passed in input stream on Node.js 4.x and below.", "recommendation": "No fix is currently available for this vulnerability. It is our recommendation to not install or use this module if user input is being passed in to \`stringstream\`.\\n", "references": "- [HackerOne Report](https://hackerone.com/reports/321670)\\n- https://github.com/mhart/StringStream/blob/v0.0.5/stringstream.js#L32", "access": "public", "severity": "moderate", "cwe": "CWE-125", "metadata": { "module_type": "", "exploitability": 2, "affected_components": "" }, "url": "https://nodesecurity.io/advisories/664" } }, "muted": [], "metadata": { "vulnerabilities": { "info": 0, "low": 23, "moderate": 6, "high": 2, "critical": 1 }, "dependencies": 0, "devDependencies": 965, "optionalDependencies": 28, "totalDependencies": 965 }, "runId": "d558cef6-7061-454e-89d3-ee85181d1dd6" }`
