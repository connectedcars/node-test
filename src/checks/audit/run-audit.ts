import { RunProcess } from '../../unix/run-process'
import { AuditData } from './audit-types'

export async function runAudit(command = 'npm', extraArgs: string[] = []): Promise<AuditData> {
  const cmd = new RunProcess(command, ['audit', '--json', ...extraArgs], {
    env: { ...process.env, TZ: 'UTC' }
  })
  const data: Buffer[] = []
  cmd.stdout?.on('data', chunk => {
    data.push(chunk)
  })
  await cmd.waitForStarted()
  await cmd.waitForExit()
  const json = Buffer.concat(data).toString('utf8')
  return JSON.parse(json)
}
