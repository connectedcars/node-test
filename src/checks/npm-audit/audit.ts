import { CheckRunResult } from '../checks-common'
import { Advisory, AuditData, Vulnerabilities } from './audit-types'

export interface AuditInput {
  data: AuditData
}

function getSummary(problems: Vulnerabilities, totalDependencies: number): string {
  let summary = `Found **${problems.all}** vulnerabilities`
  if (problems.all > 0) {
    const details = []
    for (const key in problems) {
      if (key === 'all') {
        continue
      }
      if (problems[key] > 0) {
        details.push(`**${problems[key]}** ${key}`)
      }
    }
    summary += ` (${details.join(', ')})`
  }
  summary += ` in ${totalDependencies} scanned packages`
  return summary
}

interface SeverityMap {
  [name: string]: number
}

const severities: SeverityMap = {
  critical: 5,
  high: 4,
  moderate: 3,
  low: 2,
  info: 1
}

function getText(data: AuditData): string {
  const advisories = Object.values(data.advisories)
  advisories.sort((a: Advisory, b: Advisory) => severities[b.severity] - severities[a.severity])
  const entries: string[] = []
  for (const advisory of advisories) {
    const severity = advisory.severity.substr(0, 1).toUpperCase() + advisory.severity.substr(1)
    let entry = `## ${severity}: ${advisory.title}\n`
    const overview = advisory.overview.replace(/\r/g, '')
    entry += `${overview}\n\n`
    const id = advisory.id
    const url = advisory.url
    entry += `[Read about advisory ${id} at nodesecurity.io](${url})\n\n`
    entry += `#### ${advisory.module_name} (${advisory.vulnerable_versions})\n`
    for (const finding of advisory.findings) {
      for (const path of finding.paths) {
        const dev = finding.dev ? '**DEV:**' : ''
        entry += ` - ${dev} \`${path.replace(/>/g, '` > `')}\`\n`
      }
    }
    const recommendation = advisory.recommendation.replace(/\r/g, '')
    entry += `\n#### Recommendation\n${recommendation}`
    entries.push(entry)
  }
  return entries.join('\n\n')
}

export function auditCheck({ data }: AuditInput): CheckRunResult {
  const problems = {
    all: 0,
    info: 0,
    low: 0,
    moderate: 0,
    high: 0,
    critical: 0
  }
  let totalDependencies = 0
  if (data.metadata) {
    if (data.metadata.vulnerabilities) {
      Object.assign(problems, data.metadata.vulnerabilities)
      problems.all = Object.values(problems).reduce((sum, val) => sum + val, 0)
    }
    if (data.metadata.totalDependencies) {
      totalDependencies = data.metadata.totalDependencies
    }
  }
  return {
    conclusion: problems.all === 0 ? 'success' : 'neutral',
    output: {
      title: 'npm audit security report',
      summary: getSummary(problems, totalDependencies),
      text: getText(data)
    }
  }
}
