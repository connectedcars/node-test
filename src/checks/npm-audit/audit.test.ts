import { auditCheck } from './audit'
import { auditMultiVulnerabilities, auditNoVulnerabilities } from './resources/audit-help-text'

describe('checks/audit', () => {
  it('converts successful audit', () => {
    const data = JSON.parse(auditNoVulnerabilities)
    const result = auditCheck({
      data,
      sha: 'c61a4ae014360e064eb2a9f76c8a6a55d05e5b88'
    })
    expect(result).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })

  it('converts failed audit', () => {
    const data = JSON.parse(auditMultiVulnerabilities)
    const result = auditCheck({
      sha: 'abcdefg',
      data
    })
    expect(result).toMatchSnapshot({
      completed_at: expect.stringMatching(/^\d{4}/)
    })
  })
})
