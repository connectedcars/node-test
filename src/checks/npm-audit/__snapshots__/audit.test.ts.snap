// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checks/audit converts failed audit 1`] = `
Object {
  "conclusion": "neutral",
  "output": Object {
    "summary": "Found **32** vulnerabilities (**23** low, **6** moderate, **2** high, **1** critical) in 965 scanned packages",
    "text": "## Critical: Command Injection
Affected versions of \`growl\` do not properly sanitize input prior to passing it into a shell command, allowing for arbitrary command execution.

[Read about advisory 146 at nodesecurity.io](https://nodesecurity.io/advisories/146)

#### growl (<1.10.2)
 - **DEV:** \`mocha\` > \`growl\`

#### Recommendation
Update to version 1.10.2 or later.

## High: Regular Expression Denial of Service
Affected versions of \`tough-cookie\` are susceptible to a regular expression denial of service.

The amplification on this vulnerability is relatively low - it takes around 2 seconds for the engine to execute on a malicious input which is 50,000 characters in length.

If node was compiled using the \`-DHTTP_MAX_HEADER_SIZE\` however, the impact of the vulnerability can be significant, as the primary limitation for the vulnerability is the default max HTTP header length in node.

[Read about advisory 525 at nodesecurity.io](https://nodesecurity.io/advisories/525)

#### tough-cookie (<2.3.3)
 - **DEV:** \`coveralls\` > \`request\` > \`tough-cookie\`

#### Recommendation
Update to version 2.3.3 or later.

## High: Regular Expression Denial of Service
Versions of \`sshpk\` before 1.14.1 are vulnerable to regular expression denial of service when parsing crafted invalid public keys.

[Read about advisory 606 at nodesecurity.io](https://nodesecurity.io/advisories/606)

#### sshpk (<1.14.1)
 - **DEV:** \`coveralls\` > \`request\` > \`http-signature\` > \`sshpk\`

#### Recommendation
Update to version 1.14.1 or later.

## Moderate: Prototype pollution
Versions of \`hoek\` prior to 4.2.1 and 5.0.3 are vulnerable to prototype pollution.

The \`merge\` function, and the \`applyToDefaults\` and \`applyToDefaultsWithShallow\` functions which leverage \`merge\` behind the scenes, are vulnerable to a prototype pollution attack when provided an _unvalidated_ payload created from a JSON string containing the \`__proto__\` property.

This can be demonstrated like so:

\`\`\`javascript
var Hoek = require('hoek');
var malicious_payload = '{\\"__proto__\\":{\\"oops\\":\\"It works !\\"}}';

var a = {};
console.log(\\"Before : \\" + a.oops);
Hoek.merge({}, JSON.parse(malicious_payload));
console.log(\\"After : \\" + a.oops);
\`\`\`

This type of attack can be used to overwrite existing properties causing a potential denial of service.

[Read about advisory 566 at nodesecurity.io](https://nodesecurity.io/advisories/566)

#### hoek (<= 4.2.0 || >= 5.0.0 < 5.0.3)
 - **DEV:** \`coveralls\` > \`request\` > \`hawk\` > \`boom\` > \`hoek\`
 - **DEV:** \`coveralls\` > \`request\` > \`hawk\` > \`cryptiles\` > \`boom\` > \`hoek\`
 - **DEV:** \`coveralls\` > \`request\` > \`hawk\` > \`hoek\`
 - **DEV:** \`coveralls\` > \`request\` > \`hawk\` > \`sntp\` > \`hoek\`

#### Recommendation
Update to version 4.2.1, 5.0.3 or later.

## Moderate: Memory Exposure
Versions of \`tunnel-agent\` before 0.6.0 are vulnerable to memory exposure.

This is exploitable if user supplied input is provided to the auth value and is a number.

Proof-of-concept:
\`\`\`js
require('request')({
  method: 'GET',
  uri: 'http://www.example.com',
  tunnel: true,
  proxy:{
    protocol: 'http:',
    host:'127.0.0.1',
    port:8080,
    auth:USERSUPPLIEDINPUT // number
  }
});
\`\`\`

[Read about advisory 598 at nodesecurity.io](https://nodesecurity.io/advisories/598)

#### tunnel-agent (<0.6.0)
 - **DEV:** \`coveralls\` > \`request\` > \`tunnel-agent\`

#### Recommendation
Update to version 0.6.0 or later.

## Moderate: Out-of-bounds Read
All versions of \`stringstream\` are vulnerable to out-of-bounds read as it allocates uninitialized Buffers when number is passed in input stream on Node.js 4.x and below.

[Read about advisory 664 at nodesecurity.io](https://nodesecurity.io/advisories/664)

#### stringstream (<=0.0.5)
 - **DEV:** \`coveralls\` > \`request\` > \`stringstream\`

#### Recommendation
No fix is currently available for this vulnerability. It is our recommendation to not install or use this module if user input is being passed in to \`stringstream\`.


## Low: Cryptographically Weak PRNG
Affected versions of \`randomatic\` generate random values using a cryptographically weak psuedo-random number generator. This may result in predictable values instead of random values as intended.



[Read about advisory 157 at nodesecurity.io](https://nodesecurity.io/advisories/157)

#### randomatic (<3.0.0)
 - **DEV:** \`nyc\` > \`micromatch\` > \`braces\` > \`expand-range\` > \`fill-range\` > \`randomatic\`
 - **DEV:** \`nyc\` > \`test-exclude\` > \`micromatch\` > \`braces\` > \`expand-range\` > \`fill-range\` > \`randomatic\`

#### Recommendation
Update to version 3.0.0 or later.


## Low: Regular Expression Denial of Service
Affected versions of \`debug\` are vulnerable to regular expression denial of service when untrusted user input is passed into the \`o\` formatter. 

As it takes 50,000 characters to block the event loop for 2 seconds, this issue is a low severity issue.

[Read about advisory 534 at nodesecurity.io](https://nodesecurity.io/advisories/534)

#### debug (<= 2.6.8 || >= 3.0.0 <= 3.0.1)
 - **DEV:** \`babel-eslint\` > \`babel-traverse\` > \`debug\`
 - **DEV:** \`eslint\` > \`debug\`
 - **DEV:** \`mocha\` > \`debug\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-template\` > \`babel-traverse\` > \`debug\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-traverse\` > \`debug\`
 - **DEV:** \`nyc\` > \`istanbul-lib-source-maps\` > \`debug\`

#### Recommendation
Version 2.x.x: Update to version 2.6.9 or later.
Version 3.x.x: Update to version 3.1.0 or later.


## Low: Prototype Pollution
Versions of \`lodash\` before 4.17.5 are vulnerable to prototype pollution. 

The vulnerable functions are 'defaultsDeep', 'merge', and 'mergeWith' which allow a malicious user to modify the prototype of \`Object\` via \`__proto__\` causing the addition or modification of an existing property that will exist on all objects.



[Read about advisory 577 at nodesecurity.io](https://nodesecurity.io/advisories/577)

#### lodash (<4.17.5)
 - **DEV:** \`babel-eslint\` > \`babel-traverse\` > \`babel-types\` > \`lodash\`
 - **DEV:** \`babel-eslint\` > \`babel-traverse\` > \`lodash\`
 - **DEV:** \`babel-eslint\` > \`babel-types\` > \`lodash\`
 - **DEV:** \`eslint\` > \`inquirer\` > \`lodash\`
 - **DEV:** \`eslint\` > \`lodash\`
 - **DEV:** \`eslint\` > \`table\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-generator\` > \`babel-types\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-generator\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-template\` > \`babel-traverse\` > \`babel-types\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-template\` > \`babel-traverse\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-template\` > \`babel-types\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-template\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-traverse\` > \`babel-types\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-traverse\` > \`lodash\`
 - **DEV:** \`nyc\` > \`istanbul-lib-instrument\` > \`babel-types\` > \`lodash\`

#### Recommendation
Update to version 4.17.5 or later.",
    "title": "npm audit security report",
  },
}
`;
